src/VisualStudio/IntegrationTest/TestUtilities/InProcess/InProcComponent.cs

```diff
@@ -47,29 +47,30 @@ protected JoinableTaskFactory JoinableTaskFactory
             }
         }
 
-        protected static void BeginInvokeOnUIThread(Action action)
-#pragma warning disable VSTHRD001 // Avoid legacy thread switching APIs
-            => CurrentApplicationDispatcher.BeginInvoke(action, DispatcherPriority.Background);
-#pragma warning restore VSTHRD001 // Avoid legacy thread switching APIs
-
-        protected static void InvokeOnUIThread(Action action)
+        protected static void InvokeOnUIThread(Action<CancellationToken> action)
+        {
+            using var cancellationTokenSource = new CancellationTokenSource(Helper.HangMitigatingTimeout);
 #pragma warning disable VSTHRD001 // Avoid legacy thread switching APIs
-            => CurrentApplicationDispatcher.Invoke(action, DispatcherPriority.Background);
+            CurrentApplicationDispatcher.Invoke(() => action(cancellationTokenSource.Token), DispatcherPriority.Background, cancellationToken: cancellationTokenSource.Token);
```

Comment:
To confirm intent, the intent is the cancellation exception propagates, and so the test will fail if a timeout happens? Or should the timeout simply let the test continue?

Reply:
test will fail